pyro/infer/mcmc/api.py

```diff
@@ -377,6 +374,12 @@ def model(data):
         z_flat_acc = [[] for _ in range(self.num_chains)]
         with optional(pyro.validation_enabled(not self.disable_validation),
                       self.disable_validation is not None):
+            if self.num_chains > 1:
+                # XXX we clone CUDA tensor args to resolve the issue "Invalid device pointer"
+                # at https://github.com/pytorch/pytorch/issues/10375
+                # This also resolves "RuntimeError: Cowardly refusing to serialize non-leaf tensor which
+                # requires_grad", which happens with `jit_compile` under PyTorch 1.7
+                args = [arg.clone().detach() if torch.is_tensor(arg) else arg for arg in args]
```

Comment:
@neerajprad Could you double-check if this is a good solution? This resolves the issues for:
+ cuda + jit/nojit
+ cpu + jit

There is no issue with `cpu + nojit` so should we filter it out at `if self.num_chains > 1` above?

Reply:
How about we remove the `num_chains` restriction and simply `detach` (instead of cloning)? That should be very cheap and we can do that as a sanity measure anyways. I think @fritzo has correctly identified that the jit is incorrectly propagating up `requires_grad`, so this seems like a bigger problem.

Reply:
Thanks! It seems that `.detach()` solves the issue. Removing `num_chains` helps for single-chain too (without detach, baseball failed with num_chains=1???). I think the problem comes from [this line](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/torch.py#L88) of `Binomial.log_prob`. If I change `k * self.logits` to `(k + 0) * self.logits`, the problem goes away without having to `detach`... That observation agrees with: `the jit is incorrectly propagating up requires_grad`.

Reply:
Thanks, @fehiepsi. I wonder if the problem is due to the cached logits. I have seen this previously for transforms where the cached value has its `requires_grad` set during backward, but in that case we get an error saying that JIT cannot insert a constant with a `requires_grad` attribute set. That doesn't explain why adding 0 helps take care of this.

Reply:
I'm not sure. I remember that the issue still happens when I changed `self.logits` to `self.probs`.